<a href="https://colab.research.google.com/github/agu3/ds4bmeTest/blob/master/a8_q1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fitting a prediction model for the mnist data with multiple hidden layers

In [0]:
!pip install -U tensorflow_datasets

In [0]:
import tensorflow as tf
import matplotlib.pyplot as plt
import numpy as np

Importing the MNIST dataset

In [0]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0 

11493376/11490434 [==============================] - 0s 0us/step


In [0]:
[x_train.shape, y_train.shape, x_test.shape, y_test.shape]

[(60000, 28, 28), (60000,), (10000, 28, 28), (10000,)]

Generating the model

In [0]:
model1 = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape = (28, 28)),
  tf.keras.layers.Dense(128, activation = 'relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(64, activation = 'relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(32, activation = 'relu' ),
  tf.keras.layers.Dropout(0.2), 
  tf.keras.layers.Dense(10)
])

In [0]:
model1.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 64)                8256      
_________________________________________________________________
dropout_1 (Dropout)          (None, 64)                0         
_________________________________________________________________
dense_2 (Dense)              (None, 32)                2080      
_________________________________________________________________
dropout_2 (Dropout)          (None, 32)                0

Create a loss function 

In [0]:
loss_fn1 = tf.keras.losses.SparseCategoricalCrossentropy(from_logits = True)

Giving the model an optimizer, loss function, and telling it what metrics to save 

In [0]:
model1.compile(
    optimizer = 'adam',
    loss = loss_fn1,
    metrics = ['accuracy']
)

Fit the model

In [0]:
model1.fit(x_train, y_train, epochs = 5)

Epoch 1/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.4095 - accuracy: 0.8781
Epoch 2/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1950 - accuracy: 0.9464
Epoch 3/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1547 - accuracy: 0.9572
Epoch 4/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1353 - accuracy: 0.9629
Epoch 5/5
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1162 - accuracy: 0.9669


Evaluate the model on the test data

In [0]:
model1.evaluate(x_test, y_test, verbose = 2)

313/313 - 0s - loss: 0.0876 - accuracy: 0.9758


[0.08757806569337845, 0.9757999777793884]

# Fitting a prediction model for the mnist data using just one hidden layer


In [0]:
mnist = tf.keras.datasets.mnist 

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train / 255.0
x_test = x_test / 255.0


In [0]:
[x_train.shape, y_train.shape, x_test.shape, y_test.shape]

[(60000, 28, 28), (60000,), (10000, 28, 28), (10000,)]

Generating the model with one hidden layer 

In [0]:
model2 = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape = (28, 28)),
  tf.keras.layers.Dense(32, activation = 'relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])

In [0]:
model2.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_1 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_4 (Dense)              (None, 32)                25120     
_________________________________________________________________
dropout_3 (Dropout)          (None, 32)                0         
_________________________________________________________________
dense_5 (Dense)              (None, 10)                330       
Total params: 25,450
Trainable params: 25,450
Non-trainable params: 0
_________________________________________________________________


Create a loss function and compile the model

In [0]:
loss_fn2 = tf.keras.losses.SparseCategoricalCrossentropy(from_logits = True)

model2.compile(optimizer = 'adam',
               loss = loss_fn2,
               metrics = ['accuracy']
               )

Fit the model

In [0]:
model2.fit(x_train, y_train, epochs = 5)

Epoch 1/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.4574 - accuracy: 0.8652
Epoch 2/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2775 - accuracy: 0.9176
Epoch 3/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2434 - accuracy: 0.9283
Epoch 4/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2231 - accuracy: 0.9330
Epoch 5/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.2090 - accuracy: 0.9383


Evaluate the model on test data

In [0]:
model2.evaluate(x_test, y_test, verbose = 2)

313/313 - 0s - loss: 0.1408 - accuracy: 0.9571


[0.14077797532081604, 0.957099974155426]

The test set accuracy using model1 with multiple hidden layers is 0.9758 while the test set accuracy using model2 with one hidden layer is 0.9571. The test set accuracy using multiple hidden layers is greater than the test set accuracy using one multiple hidden layer. 